ncm-spma/src/main/perl/spma/apt.pm

```diff
@@ -0,0 +1,433 @@
+Readonly::Scalar my $DIR_PREFERENCES => "/etc/apt/preferences.d";
+Readonly::Scalar my $TEMPLATE_REPOS => "spma/apt-source.tt";
+Readonly::Scalar my $TEMPLATE_PREFERENCES => "spma/apt-preferences.tt";
+Readonly::Scalar my $TREE_REPOS => "/software/repositories";
+Readonly::Scalar my $TREE_PKGS => "/software/packages";
+Readonly::Scalar my $TREE_COMPONENT => "/software/components/${project.artifactId}";
+Readonly::Scalar my $BIN_APT_GET => "/usr/bin/apt-get";
+Readonly::Scalar my $BIN_APT_MARK => "/usr/bin/apt-mark";
+Readonly::Scalar my $BIN_DPKG_QUERY => "/usr/bin/dpkg-query";
+Readonly::Scalar my $CMD_APT_UPDATE => [$BIN_APT_GET, qw(-qq update)];
+Readonly::Scalar my $CMD_APT_UPGRADE => [$BIN_APT_GET, qw(-qq dist-upgrade)];
+Readonly::Scalar my $CMD_APT_INSTALL => [$BIN_APT_GET, qw(-qq install)];
+Readonly::Scalar my $CMD_APT_AUTOREMOVE => [$BIN_APT_GET, qw(-qq autoremove)];
+Readonly::Scalar my $CMD_APT_MARK => [$BIN_APT_MARK, qw(-qq)];
+Readonly::Scalar my $CMD_DPKG_QUERY => [$BIN_DPKG_QUERY, qw(-W -f=${db:Status-Abbrev};${Package}\n)];
+
+our $NoActionSupported = 1;
+
+
+# If user packages are not allowed, removes any repositories present
+# in the system that are not listed in $allowed_repos.
+sub cleanup_old_repos
+{
+    my ($self, $repo_dir, $allowed_repos, $allow_user_pkgs) = @_;
+    $self->debug(5, 'Entered cleanup_old_repos()');
+
+    return 1 if $allow_user_pkgs;
+
+    my $dir;
+    if (!opendir($dir, $repo_dir)) {
+        $self->error("Unable to read repositories in $repo_dir");
+        return 0;
+    }
+
+    my $current = Set::Scalar->new(map(m{(.*)\.list$}, readdir($dir)));
+
+    closedir($dir);
+    my $allowed = Set::Scalar->new(map($_->{name}, @$allowed_repos));
+    my $rm = $current-$allowed;
+    foreach my $i (@$rm) {
+        # We use $f here to make Devel::Cover happy
+        my $f = "$repo_dir/$i.list";
+        $self->verbose("Unlinking outdated repository $f");
+        if (!unlink($f)) {
+            $self->error("Unable to remove outdated repository $i: $!");
+            return 0;
+        }
+    }
+    return 1;
+}
+
+
+# Creates the repository dir if needed.
+sub initialize_repos_dir
+{
+    my ($self, $repo_dir) = @_;
+    $self->debug(5, 'Entered initialize_repos_dir()');
+
+    if (! -d $repo_dir) {
+        $self->verbose("$repo_dir didn't exist. Creating it");
+        if (!eval{mkpath($repo_dir)} || $@) {
+            $self->error("Unable to create repository dir $repo_dir: $@");
+            return 0;
+        }
+    }
+    return 1;
+}
+
+
+# Returns the URLs of the repositories listed in $prots, with their
+# host part replaced by each of the reverse @proxies.
+sub generate_reverse_proxy_urls
+{
+    my ($self, $prots, @proxies) = @_;
+    $self->debug(5, 'Entered generate_reverse_proxy_urls()');
+
+    my @l;
+    foreach my $pt (@$prots) {
+        foreach my $px (@proxies) {
+            my $url = $pt->{url};
+            $url =~ s{^(.*?):(/+)[^/]+}{$1:$2$px};
```

Comment:
should you handle the case that this regex doesn't match anything? it also looks like a confusiong way to write 
```perl
push(@l, ($pt->{url} =~ m/^(.*?):(/+)/) ? "$1:$2$px" : $pt->{url})
```